the-collab-lab/tcl-33-smart-shopping-list
# Added a search bar on View List page to filter grocery list items

## Description

We added a search bar that allows the user to filter through the existing grocery items. The search bar also provides a delete button to clear the search. Each character the user enters is matched against the existing grocery items; if an item contains that character or specific sequence of characters, the item is rendered to the page.

The fastest and easiest way to add the delete button that cleared the search was available by using the type="search" attribute.

Our team encountered some unexpected difficulties with page loading that required additional conditionals to account for empty and non-existent lists. 

## Related Issue

Closes #9

## Acceptance Criteria

- Display a text field above the top of the shopping list
- As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
- When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
- The filter text should match any part of the item name (i.e. it should not only match from the start of the string)

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |


### Before



![Screenshot (411)_LI](https://user-images.githubusercontent.com/50176611/141032025-9e885f35-fb0e-4ad9-a96e-4b090ef004cc.jpg)

### After

![Screenshot (408)](https://user-images.githubusercontent.com/50176611/141031653-ad16fb6d-7628-4d71-a3eb-b55b6ad59bc5.png)

![filteredList](https://user-images.githubusercontent.com/50176611/141031283-bc731a35-3e74-4a2e-a43c-dd401888b7db.gif)

## Testing Steps / QA Criteria

- Go to the main branch of tcl-33-smart-shopping-list
- Git checkout is-jt-filter-list 
- npm start
- When the app is open in your browser, if you have a token saved in your local storage, click on the "View List" tab in the navbar and it should take you to the View List page
- If you do not have a token saved, start a new list, add a couple of items, and go to the View List page to see your list
- In the search bar, enter a character of a word in one of your grocery items. For instance, if one of your grocery items is "pumpkin," enter "k" into the search bar. You do not have to worry about matching uppercase/lowercase--the search function is case insensitive!
- All grocery items containing the word "k" will appear under the search bar. If you add additional characters, as long as they are in the sequence of at least one of the grocery items, those items will appear in the list.
- To clear the list, simply click on the "X" in the search bar and it will immediately clear your entry.
